src/bower_components/apiclient/connectionmanager.js

```diff
@@ -239,307 +396,543 @@ define(["events", "apiclient", "appStorage"], function(events, apiClientFactory,
+            return new Promise(function (resolve, reject) {
+                var state = {};
+                state.numAddresses = addresses.length;
+                state.rejects = 0;
+                addresses.map(function (url) {
+                    setTimeout(function () {
+                        if (!state.resolved) {
+                            getTryConnectPromise(url.url, url.mode, state, resolve, reject);
+                        }
+                    }, url.timeout);
                 });
-                updateServerInfo(server, systemInfo), server.LastConnectionMode = connectionMode, !1 !== options.updateDateLastAccessed && (server.DateLastAccessed = (new Date).getTime()), credentialProvider.addOrUpdateServer(credentials.Servers, server), credentialProvider.credentials(credentials);
-                var result = {
-                    Servers: []
-                };
-                result.ApiClient = self._getOrAddApiClient(server, serverUrl), result.ApiClient.setSystemInfo(systemInfo), result.State = server.AccessToken && !1 !== options.enableAutoLogin ? "SignedIn" : "ServerSignIn", result.Servers.push(server), result.ApiClient.enableAutomaticBitrateDetection = options.enableAutomaticBitrateDetection, result.ApiClient.updateServerInfo(server, serverUrl);
-                var resolveActions = function() {
-                    resolve(result), events.trigger(self, "connected", [result])
-                };
-                "SignedIn" === result.State ? (afterConnected(result.ApiClient, options), result.ApiClient.getCurrentUser().then(function(user) {
-                    onLocalUserSignIn(server, serverUrl, user).then(resolveActions, resolveActions)
-                }, resolveActions)) : resolveActions()
+            });
+        }
+
+        function onSuccessfulConnection(server, systemInfo, connectionMode, serverUrl, options, resolve) {
+            var credentials = credentialProvider.credentials();
+            options = options || {};
+
+            if (credentials.ConnectAccessToken && false !== options.enableAutoLogin) {
+                ensureConnectUser(credentials).then(function () {
+                    if (server.ExchangeToken) {
+                        addAuthenticationInfoFromConnect(server, serverUrl, credentials).then(function () {
+                            afterConnectValidated(server, credentials, systemInfo, connectionMode, serverUrl, true, options, resolve);
+                        }, function () {
+                            afterConnectValidated(server, credentials, systemInfo, connectionMode, serverUrl, true, options, resolve);
+                        });
+                    } else {
+                        afterConnectValidated(server, credentials, systemInfo, connectionMode, serverUrl, true, options, resolve);
+                    }
+                });
+            } else {
+                afterConnectValidated(server, credentials, systemInfo, connectionMode, serverUrl, true, options, resolve);
```

Comment:
I wonder what is going on here... this `afterConnectValidated()` seems to be called in 4 places here with exact same parameters